core/src/message.rs

```diff
@@ -13,40 +13,111 @@
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-//! Transaction related types used in the Radicle Registry.
-
+//! All transaction messages that can be submitted to the ledger
+//!
+//! See the README.md for more information on how to document messages.
 extern crate alloc;
 
 use crate::{AccountId, Balance, Bytes128, CheckpointId, ProjectId};
 use parity_scale_codec::{Decode, Encode};
 use sp_core::H256;
 
+/// Registers a project on the Radicle Registry with the given ID.
+///
+/// # State changes
+///
+/// If successful, a new [crate::Project] with the given properties is added to the state.
+///
+/// [crate::Project::members] is initialized with the transaction author.
+///
+/// [crate::Project::account_id] is generated randomly.
+///
+/// # State-dependent validations
+///
+/// A project with the same ID must not yet exist.
+///
+/// A checkpoint with the given ID must exist.
 #[derive(Decode, Encode, Clone, Debug, Eq, PartialEq)]
 pub struct RegisterProject {
     pub id: ProjectId,
+
+    /// Initial checkpoint of the project.
     pub checkpoint_id: CheckpointId,
+
+    /// Opaque metadata that cannot be changed.
+    ///
+    /// Used by the application.
     pub metadata: Bytes128,
 }
 
+/// Add a new checkpoint to the state.
+///
+/// # State changes
+///
+/// If successful, adds a new [crate::Checkpoint] with the given parameters to the state.
+///
+/// # State-dependent validations
+///
+/// If `previous_checkpoint_id` is provided a checkpoint with the given ID must exist in the state.
```

Comment:
```suggestion
/// The `previous_checkpoint_id` must exist in the state.
```

Reply:
I’d like to keep the “if” because `previous_checkpoint_id` is an `Option`. I’d also prefer to be explicit about distinguishing IDs and their existence in the state.

Reply:
Ok. In that case, a comma would help:

```suggestion
/// If `previous_checkpoint_id` is provided, a checkpoint with the given ID must exist in the state.
```